File: app-macos/Sources/RunnerForge/Services/RunnerSupervisor.swift
import Foundation

public enum ReplicaState: String, Sendable {
    case stopped, starting, waitingForJob, runningJob, cleaningUp, error

    public var label: String {
        switch self {
        case .stopped: "Stopped"
        case .starting: "Starting"
        case .waitingForJob: "Waiting for job"
        case .runningJob: "Running job"
        case .cleaningUp: "Cleaning up"
        case .error: "Error"
        }
    }
}

public struct ReplicaStatus: Sendable, Identifiable, Hashable {
    public let id = UUID()
    public var classId: String
    public var index: Int
    public var runnerName: String = ""
    public var state: ReplicaState = .stopped
    public var startedAt: Date?
    public var jobsCompletedToday: Int = 0
    public var lastError: String?
}

/// Crash-safe record of what was running, written before launch.
struct SupervisorState: Codable {
    var runners: [SupervisorRunnerEntry] = []
}

struct SupervisorRunnerEntry: Codable {
    var runnerName: String
    var classId: String
    var pid: Int32?
    var startedAt: Date
    var repo: String
}

/// Starts, watches and replaces ephemeral runners.
///
/// Every runner takes ONE job and dies. State is written to
/// `paths.workDir/state.json` BEFORE a process is launched, so a hard power cut
/// leaves a record the next launch can clean up rather than an orphan nobody
/// knows about.
public actor RunnerSupervisor {
    private let logBus: LogBus
    private let tartService: TartService
    private let appService: GitHubAppService
    private let actionsService: GitHubActionsService
    private let reaperService: ReaperService

    private var replicas: [String: ReplicaStatus] = [:]
    private var tasks: [String: Task<Void, Never>] = [:]
    private var recentFailures: [String: [Date]] = [:]

    /// Three non-zero exits within five minutes stops the class. A broken class —
    /// bad image, wrong labels, revoked key — must never hot-loop against the
    /// GitHub API.
    public static let failureThreshold = 3
    public static let failureWindow: TimeInterval = 300

    /// Set when the Reaper reported survivors. Blocks new runners until acknowledged.
    public private(set) var blockedByStraySurvivors = false
    public private(set) var straySurvivorDetails: [String] = []

    public var scriptsDirectory: String = "/usr/local/share/runnerforge/scripts"

    public init(
        logBus: LogBus, tartService: TartService, appService: GitHubAppService,
        actionsService: GitHubActionsService, reaperService: ReaperService
    ) {
        self.logBus = logBus
        self.tartService = tartService
        self.appService = appService
        self.actionsService = actionsService
        self.reaperService = reaperService
    }

    public func currentReplicas() -> [ReplicaStatus] {
        replicas.values.sorted { ($0.classId, $0.index) < ($1.classId, $1.index) }
    }

    public func acknowledgeStraySurvivors() {
        blockedByStraySurvivors = false
        straySurvivorDetails = []
    }

    private func key(_ classId: String, _ index: Int) -> String { "\(classId)#\(index)" }

    public func startClass(config: ForgeConfig, runnerClass: RunnerClass) async {
        guard !blockedByStraySurvivors else {
            await logBus.error("supervisor",
                "refusing to start runners: the Reaper reported survivors that must be acknowledged first.")
            return
        }

        guard runnerClass.runsOnMac else {
            // Windows classes run on the PC. This app never virtualizes Windows
            // containers, and saying so beats a button that silently does nothing.
            await logBus.warning("supervisor",
                "\(runnerClass.classId) runs on the Windows host, not on this Mac.")
            return
        }

        guard let runnerConfig = config.runners.first(where: { $0.classId == runnerClass.classId }),
              runnerConfig.enabled else {
            await logBus.warning("supervisor", "\(runnerClass.classId) is not enabled")
            return
        }

        let count = runnerClass.replicasAreFixed ? 1 : runnerConfig.replicas

        for index in 0..<count {
            let replicaKey = key(runnerClass.classId, index)
            if let existing = replicas[replicaKey], existing.state != .stopped, existing.state != .error {
                continue
            }

            replicas[replicaKey] = ReplicaStatus(classId: runnerClass.classId, index: index)

            tasks[replicaKey] = Task { [weak self] in
                await self?.runReplicaLoop(config: config, runnerClass: runnerClass, key: replicaKey)
            }
        }
    }

    /// One replica's whole life: mint a JIT config, run one job, reap, replace.
    private func runReplicaLoop(config: ForgeConfig, runnerClass: RunnerClass, key replicaKey: String) async {
        let repo = config.github.repos.first ?? ""

        while !Task.isCancelled {
            if isBackingOff(runnerClass.classId) {
                replicas[replicaKey]?.state = .error
                replicas[replicaKey]?.lastError =
                    "stopped after \(Self.failureThreshold) failures within 5 minutes"
                await logBus.error("supervisor",
                    "\(runnerClass.classId): stopped after repeated failures. Not retrying — a broken class "
                    + "must not hot-loop.")
                return
            }

            do {
                replicas[replicaKey]?.state = .starting

                let jit = try await appService.generateJitConfig(
                    config: config, repo: repo, runnerClass: runnerClass)

                replicas[replicaKey]?.runnerName = jit.runnerName
                replicas[replicaKey]?.startedAt = Date()
                replicas[replicaKey]?.state = .waitingForJob

                // Recorded BEFORE launch: a crash between here and exit leaves a
                // record the next start can reap.
                recordState(config: config, entry: SupervisorRunnerEntry(
                    runnerName: jit.runnerName, classId: runnerClass.classId,
                    pid: nil, startedAt: Date(), repo: repo))

                let timeout = config.runners
                    .first { $0.classId == runnerClass.classId }?.jobTimeoutMinutes ?? 90

                let result = try await tartService.runEphemeralJob(
                    scriptsDirectory: scriptsDirectory,
                    image: "runnerforge-macos:1.0.0",
                    jitConfig: jit.encodedConfig,
                    cacheDirectory: (config.paths.workDir as NSString).appendingPathComponent("cache"),
                    jobTimeoutMinutes: timeout)

                replicas[replicaKey]?.state = .cleaningUp
                forgetState(config: config, runnerName: jit.runnerName)

                if result.succeeded {
                    replicas[replicaKey]?.jobsCompletedToday += 1
                    recentFailures[runnerClass.classId] = []
                } else {
                    recordFailure(runnerClass.classId)
                    replicas[replicaKey]?.lastError = "exit code \(result.exitCode)"
                    await logBus.warning("supervisor",
                        "\(jit.runnerName) exited with \(result.exitCode)")
                }

                await reapScoped(config: config)
            } catch is CancellationError {
                break
            } catch {
                recordFailure(runnerClass.classId)
                replicas[replicaKey]?.state = .error
                replicas[replicaKey]?.lastError = String(describing: error)
                await logBus.error("supervisor", String(describing: error))
                try? await Task.sleep(for: .seconds(5))
            }
        }

        replicas[replicaKey]?.state = .stopped
        replicas[replicaKey]?.startedAt = nil
    }

    private func reapScoped(config: ForgeConfig) async {
        let liveNames = Set(replicas.values
            .filter { $0.state == .runningJob || $0.state == .waitingForJob }
            .map(\.runnerName)
            .filter { !$0.isEmpty })

        guard let result = try? await reaperService.reap(
            config: config, livePids: [], liveRunnerNames: liveNames) else { return }

        if result == .straysSurvived {
            blockedByStraySurvivors = true
            let survivors = (try? await reaperService.findStrays(
                config: config, livePids: [], liveRunnerNames: liveNames)) ?? []
            straySurvivorDetails = survivors.map(\.detail)
        }
    }

    public func stopClass(_ runnerClass: RunnerClass) {
        for (replicaKey, task) in tasks where replicaKey.hasPrefix(runnerClass.classId + "#") {
            task.cancel()
            tasks[replicaKey] = nil
        }
    }

    /// Refuses new jobs and waits for in-flight ones, then forces. Called on app
    /// close, so the user never loses a running build to a window close.
    public func drain(timeout: TimeInterval, onProgress: (@Sendable (String) -> Void)? = nil) async {
        onProgress?("refusing new jobs")
        for task in tasks.values { task.cancel() }

        let deadline = Date().addingTimeInterval(timeout)

        while Date() < deadline {
            let inFlight = replicas.values.count { $0.state == .runningJob }
            if inFlight == 0 { break }
            onProgress?("waiting for \(inFlight) in-flight job(s)")
            try? await Task.sleep(for: .seconds(2))
        }

        onProgress?("drained")
    }

    /// On start, reap anything recorded but not alive, and remove any offline
    /// forge-* registration left dangling on GitHub by a crash.
    public func recoverFromCrash(config: ForgeConfig) async {
        let state = readState(config: config)

        if !state.runners.isEmpty {
            await logBus.warning("supervisor",
                "state.json records \(state.runners.count) runner(s) from a previous session; reaping")
            _ = try? await reaperService.reap(config: config)
            writeState(config: config, state: SupervisorState())
        }

        for repo in config.github.repos {
            guard let runners = try? await actionsService.listSelfHostedRunners(
                config: config, repo: repo) else { continue }

            for runner in runners where runner.name.hasPrefix("forge-") && runner.status == "offline" {
                await logBus.info("supervisor", "removing offline registration \(runner.name) from \(repo)")
                _ = try? await actionsService.deleteRunner(config: config, repo: repo, runnerId: runner.id)
            }
        }
    }

    // --- backoff -------------------------------------------------------------
    private func isBackingOff(_ classId: String) -> Bool {
        guard var failures = recentFailures[classId] else { return false }
        failures.removeAll { Date().timeIntervalSince($0) > Self.failureWindow }
        recentFailures[classId] = failures
        return failures.count >= Self.failureThreshold
    }

    private func recordFailure(_ classId: String) {
        recentFailures[classId, default: []].append(Date())
    }

    // --- crash-safe state ----------------------------------------------------
    private func statePath(config: ForgeConfig) -> String {
        (config.paths.workDir as NSString).appendingPathComponent("state.json")
    }

    private func readState(config: ForgeConfig) -> SupervisorState {
        let path = statePath(config: config)
        guard let data = FileManager.default.contents(atPath: path),
              let state = try? JSONDecoder().decode(SupervisorState.self, from: data) else {
            return SupervisorState()
        }
        return state
    }

    private func writeState(config: ForgeConfig, state: SupervisorState) {
        try? FileManager.default.createDirectory(
            atPath: config.paths.workDir, withIntermediateDirectories: true)
        guard let data = try? JSONEncoder().encode(state) else { return }
        try? data.write(to: URL(fileURLWithPath: statePath(config: config)))
    }

    private func recordState(config: ForgeConfig, entry: SupervisorRunnerEntry) {
        var state = readState(config: config)
        state.runners.append(entry)
        writeState(config: config, state: state)
    }

    private func forgetState(config: ForgeConfig, runnerName: String) {
        var state = readState(config: config)
        state.runners.removeAll { $0.runnerName == runnerName }
        writeState(config: config, state: state)
    }
}

File: app-windows/RunnerForge/Services/RunnerSupervisor.cs
using System.Collections.Concurrent;
using System.Text.Json;
using System.Text.Json.Serialization;
using RunnerForge.Models;

namespace RunnerForge.Services;

public enum ReplicaState { Stopped, Starting, WaitingForJob, RunningJob, CleaningUp, Error }

/// <summary>One live runner, as the Runners page sees it.</summary>
public sealed class ReplicaStatus
{
    public required string ClassId { get; init; }
    public required int Index { get; init; }
    public string RunnerName { get; set; } = "";
    public ReplicaState State { get; set; } = ReplicaState.Stopped;
    public DateTimeOffset? StartedAt { get; set; }
    public string? CurrentJobId { get; set; }
    public int JobsCompletedToday { get; set; }
    public string? LastError { get; set; }

    public TimeSpan Uptime => StartedAt is null ? TimeSpan.Zero : DateTimeOffset.UtcNow - StartedAt.Value;
}

/// <summary>Crash-safe record of what was running, written before launch.</summary>
public sealed class SupervisorState
{
    [JsonPropertyName("runners")] public List<SupervisorRunnerEntry> Runners { get; set; } = [];
}

public sealed class SupervisorRunnerEntry
{
    [JsonPropertyName("runnerName")] public string RunnerName { get; set; } = "";
    [JsonPropertyName("classId")] public string ClassId { get; set; } = "";
    [JsonPropertyName("containerId")] public string? ContainerId { get; set; }
    [JsonPropertyName("pid")] public int? Pid { get; set; }
    [JsonPropertyName("startedAt")] public DateTimeOffset StartedAt { get; set; }
    [JsonPropertyName("repo")] public string Repo { get; set; } = "";
}

/// <summary>
/// Starts, watches and replaces ephemeral runners.
/// </summary>
/// <remarks>
/// Every runner takes ONE job and dies. State is written to
/// <c>paths.workDir/state.json</c> BEFORE a process is launched, so a hard power
/// cut leaves a record the next launch can clean up rather than an orphan nobody
/// knows about.
/// </remarks>
public sealed class RunnerSupervisor(
    LogBus logBus,
    DockerService dockerService,
    GitHubAppService appService,
    GitHubActionsService actionsService,
    ReaperService reaperService)
{
    private readonly LogBus _logBus = logBus;
    private readonly DockerService _dockerService = dockerService;
    private readonly GitHubAppService _appService = appService;
    private readonly GitHubActionsService _actionsService = actionsService;
    private readonly ReaperService _reaperService = reaperService;

    private readonly ConcurrentDictionary<string, ReplicaStatus> _replicas = new();
    private readonly ConcurrentDictionary<string, CancellationTokenSource> _cancellations = new();
    private readonly ConcurrentDictionary<string, List<DateTimeOffset>> _recentFailures = new();
    private readonly Lock _stateLock = new();

    /// <summary>
    /// Three non-zero exits within five minutes stops the class. A broken class —
    /// bad image, wrong labels, revoked key — must never hot-loop against the
    /// GitHub API.
    /// </summary>
    public const int FailureThreshold = 3;
    public static readonly TimeSpan FailureWindow = TimeSpan.FromMinutes(5);

    public event Action? StatusChanged;

    public IReadOnlyCollection<ReplicaStatus> Replicas => [.. _replicas.Values];

    /// <summary>Set when the Reaper reported survivors. Blocks new runners until acknowledged.</summary>
    public bool BlockedByStraySurvivors { get; private set; }

    public IReadOnlyList<string> StraySurvivorDetails { get; private set; } = [];

    public void AcknowledgeStraySurvivors()
    {
        BlockedByStraySurvivors = false;
        StraySurvivorDetails = [];
        StatusChanged?.Invoke();
    }

    private static string KeyFor(string classId, int index) => $"{classId}#{index}";

    public async Task StartClassAsync(
        ForgeConfig config, RunnerClass runnerClass, CancellationToken cancellationToken = default)
    {
        if (BlockedByStraySurvivors)
        {
            _logBus.Error("supervisor",
                "refusing to start runners: the Reaper reported survivors that must be acknowledged first.");
            return;
        }

        RunnerConfig? runnerConfig = config.Runners.FirstOrDefault(r => r.ClassId == runnerClass.ClassId);
        if (runnerConfig is null || !runnerConfig.Enabled)
        {
            _logBus.Warning("supervisor", $"{runnerClass.ClassId} is not enabled");
            return;
        }

        int replicas = runnerClass.ReplicasAreFixed ? 1 : runnerConfig.Replicas;

        for (int index = 0; index < replicas; index++)
        {
            string key = KeyFor(runnerClass.ClassId, index);
            if (_replicas.TryGetValue(key, out ReplicaStatus? existing)
                && existing.State is not (ReplicaState.Stopped or ReplicaState.Error))
            {
                continue;
            }

            var status = new ReplicaStatus { ClassId = runnerClass.ClassId, Index = index };
            _replicas[key] = status;

            var cancellation = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken);
            _cancellations[key] = cancellation;

            _ = Task.Run(() => RunReplicaLoopAsync(config, runnerClass, status, key, cancellation.Token),
                         cancellation.Token);
        }

        StatusChanged?.Invoke();
        await Task.CompletedTask.ConfigureAwait(false);
    }

    /// <summary>
    /// One replica's whole life: mint a JIT config, run one job, reap, replace.
    /// </summary>
    private async Task RunReplicaLoopAsync(
        ForgeConfig config, RunnerClass runnerClass, ReplicaStatus status, string key,
        CancellationToken cancellationToken)
    {
        string repo = config.GitHub.Repos.FirstOrDefault() ?? "";

        while (!cancellationToken.IsCancellationRequested)
        {
            if (IsBackingOff(runnerClass.ClassId))
            {
                status.State = ReplicaState.Error;
                status.LastError = $"stopped after {FailureThreshold} failures within "
                                   + $"{FailureWindow.TotalMinutes:0} minutes";
                _logBus.Error("supervisor",
                    $"{runnerClass.ClassId}: {status.LastError}. Not retrying — a broken class must not hot-loop.");
                StatusChanged?.Invoke();
                return;
            }

            try
            {
                status.State = ReplicaState.Starting;
                StatusChanged?.Invoke();

                JitConfig jit = await _appService
                    .GenerateJitConfigAsync(config, repo, runnerClass, cancellationToken).ConfigureAwait(false);

                status.RunnerName = jit.RunnerName;
                status.StartedAt = DateTimeOffset.UtcNow;
                status.State = ReplicaState.WaitingForJob;
                StatusChanged?.Invoke();

                // Recorded BEFORE launch: a crash between here and exit leaves a
                // record the next start can reap.
                RecordState(config, new SupervisorRunnerEntry
                {
                    RunnerName = jit.RunnerName,
                    ClassId = runnerClass.ClassId,
                    StartedAt = DateTimeOffset.UtcNow,
                    Repo = repo,
                });

                ProcessResult result = await StartRunnerProcessAsync(
                    config, runnerClass, jit, cancellationToken).ConfigureAwait(false);

                status.State = ReplicaState.CleaningUp;
                StatusChanged?.Invoke();

                ForgetState(config, jit.RunnerName);

                if (result.ExitCode != 0)
                {
                    RecordFailure(runnerClass.ClassId);
                    status.LastError = $"exit code {result.ExitCode}";
                    _logBus.Warning("supervisor",
                        $"{jit.RunnerName} exited with {result.ExitCode}");
                }
                else
                {
                    status.JobsCompletedToday++;
                    ClearFailures(runnerClass.ClassId);
                }

                await ReapScopedAsync(config, cancellationToken).ConfigureAwait(false);
            }
            catch (OperationCanceledException)
            {
                break;
            }
            catch (Exception ex)
            {
                RecordFailure(runnerClass.ClassId);
                status.State = ReplicaState.Error;
                status.LastError = ex.Message;
                _logBus.Error("supervisor", $"{runnerClass.ClassId} replica {status.Index}: {ex.Message}");
                StatusChanged?.Invoke();
                await Task.Delay(TimeSpan.FromSeconds(5), cancellationToken).ConfigureAwait(false);
            }
        }

        status.State = ReplicaState.Stopped;
        status.StartedAt = null;
        StatusChanged?.Invoke();
    }

    private async Task<ProcessResult> StartRunnerProcessAsync(
        ForgeConfig config, RunnerClass runnerClass, JitConfig jit, CancellationToken cancellationToken)
    {
        var mounts = new Dictionary<string, string>();

        switch (runnerClass.Isolation)
        {
            case RunnerIsolation.WindowsContainer:
                mounts["forge-fetchcontent"] = @"C:\cache\fetchcontent";
                mounts["forge-sccache"] = @"C:\cache\sccache";
                mounts["forge-aax-sdk"] = @"C:\sdk\aax";
                return await _dockerService.RunEphemeralAsync(
                    "runnerforge/win-build:1.0.0", jit.RunnerName, jit, mounts, "4", "8g", cancellationToken)
                    .ConfigureAwait(false);

            case RunnerIsolation.LinuxContainer:
                mounts["forge-fetchcontent"] = "/cache/fetchcontent";
                mounts["forge-ccache"] = "/cache/ccache";
                mounts["forge-aax-sdk"] = "/sdk/aax";
                return await _dockerService.RunEphemeralAsync(
                    "runnerforge/linux-util:1.0.0", jit.RunnerName, jit, mounts, "4", "8g", cancellationToken)
                    .ConfigureAwait(false);

            case RunnerIsolation.TartVmClone:
                // macOS runners live on the Mac. This app never virtualizes macOS:
                // it is impossible on this hardware and against Apple's licence.
                throw new PlatformNotSupportedException(
                    "mac-build runs on the Apple Silicon host, not on Windows. macOS cannot be virtualized "
                    + "on non-Apple hardware — not in Docker, QEMU, KVM or WSL.");

            case RunnerIsolation.HostProcess:
                throw new NotSupportedException(
                    "The iLok classes are started by the signing workflow on the machine holding the dongle, "
                    + "not by the supervisor.");

            default:
                throw new ArgumentOutOfRangeException(nameof(runnerClass));
        }
    }

    private async Task ReapScopedAsync(ForgeConfig config, CancellationToken cancellationToken)
    {
        int[] livePids = [.. _replicas.Values
            .Where(r => r.State is ReplicaState.RunningJob or ReplicaState.WaitingForJob)
            .Select(_ => Environment.ProcessId)];

        string[] liveNames = [.. _replicas.Values
            .Where(r => !string.IsNullOrEmpty(r.RunnerName)
                        && r.State is ReplicaState.RunningJob or ReplicaState.WaitingForJob)
            .Select(r => r.RunnerName)];

        ReaperResult result = await _reaperService
            .ReapAsync(config, livePids, liveNames, cancellationToken).ConfigureAwait(false);

        if (result == ReaperResult.StraysSurvived)
        {
            BlockedByStraySurvivors = true;
            IReadOnlyList<Stray> survivors = await _reaperService
                .FindStraysAsync(config, livePids, liveNames, cancellationToken).ConfigureAwait(false);
            StraySurvivorDetails = [.. survivors.Select(s => s.Detail)];
            StatusChanged?.Invoke();
        }
    }

    public async Task StopClassAsync(RunnerClass runnerClass)
    {
        foreach ((string key, CancellationTokenSource cancellation) in _cancellations)
        {
            if (!key.StartsWith(runnerClass.ClassId + "#", StringComparison.Ordinal)) continue;
            await cancellation.CancelAsync().ConfigureAwait(false);
        }
        StatusChanged?.Invoke();
    }

    /// <summary>
    /// Refuses new jobs and waits for in-flight ones, then forces. Called on app
    /// close, so the user never loses a running build to a window close.
    /// </summary>
    public async Task DrainAsync(TimeSpan timeout, IProgress<string>? progress = null)
    {
        progress?.Report("refusing new jobs");
        foreach (CancellationTokenSource cancellation in _cancellations.Values)
        {
            await cancellation.CancelAsync().ConfigureAwait(false);
        }

        DateTimeOffset deadline = DateTimeOffset.UtcNow + timeout;

        while (DateTimeOffset.UtcNow < deadline)
        {
            int inFlight = _replicas.Values.Count(r => r.State == ReplicaState.RunningJob);
            if (inFlight == 0) break;
            progress?.Report($"waiting for {inFlight} in-flight job(s)");
            await Task.Delay(TimeSpan.FromSeconds(2)).ConfigureAwait(false);
        }

        progress?.Report("drained");
    }

    /// <summary>
    /// On start, reap anything recorded but not alive, and remove any offline
    /// forge-* registration left dangling on GitHub by a crash.
    /// </summary>
    public async Task RecoverFromCrashAsync(ForgeConfig config, CancellationToken cancellationToken = default)
    {
        SupervisorState state = ReadState(config);
        if (state.Runners.Count > 0)
        {
            _logBus.Warning("supervisor",
                $"state.json records {state.Runners.Count} runner(s) from a previous session; reaping");
            await _reaperService.ReapAsync(config, [], [], cancellationToken).ConfigureAwait(false);
            WriteState(config, new SupervisorState());
        }

        foreach (string repo in config.GitHub.Repos)
        {
            try
            {
                IReadOnlyList<(long Id, string Name, string Status)> runners = await _actionsService
                    .ListSelfHostedRunnersAsync(config, repo, cancellationToken).ConfigureAwait(false);

                foreach ((long id, string name, string status) in runners)
                {
                    if (!name.StartsWith("forge-", StringComparison.Ordinal)) continue;
                    if (!string.Equals(status, "offline", StringComparison.OrdinalIgnoreCase)) continue;

                    _logBus.Info("supervisor", $"removing offline registration {name} from {repo}");
                    await _actionsService.DeleteRunnerAsync(config, repo, id, cancellationToken)
                        .ConfigureAwait(false);
                }
            }
            catch (Exception ex)
            {
                _logBus.Warning("supervisor", $"crash recovery could not query {repo}: {ex.Message}");
            }
        }
    }

    // --- backoff -----------------------------------------------------------
    private bool IsBackingOff(string classId)
    {
        if (!_recentFailures.TryGetValue(classId, out List<DateTimeOffset>? failures)) return false;
        lock (_stateLock)
        {
            failures.RemoveAll(f => DateTimeOffset.UtcNow - f > FailureWindow);
            return failures.Count >= FailureThreshold;
        }
    }

    private void RecordFailure(string classId)
    {
        List<DateTimeOffset> failures = _recentFailures.GetOrAdd(classId, _ => []);
        lock (_stateLock) failures.Add(DateTimeOffset.UtcNow);
    }

    private void ClearFailures(string classId)
    {
        if (_recentFailures.TryGetValue(classId, out List<DateTimeOffset>? failures))
        {
            lock (_stateLock) failures.Clear();
        }
    }

    // --- crash-safe state --------------------------------------------------
    private static string StatePath(ForgeConfig config) => Path.Combine(config.Paths.WorkDir, "state.json");

    private SupervisorState ReadState(ForgeConfig config)
    {
        string path = StatePath(config);
        if (!File.Exists(path)) return new SupervisorState();
        try
        {
            return JsonSerializer.Deserialize<SupervisorState>(File.ReadAllText(path)) ?? new SupervisorState();
        }
        catch (Exception ex)
        {
            _logBus.Warning("supervisor", $"could not read state.json: {ex.Message}");
            return new SupervisorState();
        }
    }

    private void WriteState(ForgeConfig config, SupervisorState state)
    {
        try
        {
            Directory.CreateDirectory(config.Paths.WorkDir);
            File.WriteAllText(StatePath(config),
                JsonSerializer.Serialize(state, ForgeConfig.SerializerOptions));
        }
        catch (Exception ex)
        {
            _logBus.Warning("supervisor", $"could not write state.json: {ex.Message}");
        }
    }

    private void RecordState(ForgeConfig config, SupervisorRunnerEntry entry)
    {
        lock (_stateLock)
        {
            SupervisorState state = ReadState(config);
            state.Runners.Add(entry);
            WriteState(config, state);
        }
    }

    private void ForgetState(ForgeConfig config, string runnerName)
    {
        lock (_stateLock)
        {
            SupervisorState state = ReadState(config);
            state.Runners.RemoveAll(r => r.RunnerName == runnerName);
            WriteState(config, state);
        }
    }
}
